pkg/storage/jetstream/jetstream.go

```diff
@@ -0,0 +1,134 @@
+package jetstream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/iancoleman/strcase"
+	"github.com/nats-io/nats.go"
+	corev1 "github.com/rancher/opni/pkg/apis/core/v1"
+	"github.com/rancher/opni/pkg/config/v1beta1"
+	"github.com/rancher/opni/pkg/logger"
+	"github.com/rancher/opni/pkg/storage"
+	"go.uber.org/zap"
+)
+
+const (
+	tokensBucket       = "tokens"
+	clustersBucket     = "clusters"
+	keyringsBucket     = "keyrings"
+	rolesBucket        = "roles"
+	roleBindingsBucket = "rolebindings"
+	dynamicBucket      = "dynamic"
+)
+
+type JetStreamStore struct {
+	JetStreamStoreOptions
+
+	nc *nats.Conn
+	js nats.JetStreamContext
+
+	kv struct {
+		Tokens       nats.KeyValue
+		Clusters     nats.KeyValue
+		Keyrings     nats.KeyValue
+		Roles        nats.KeyValue
+		RoleBindings nats.KeyValue
+	}
+	logger *zap.SugaredLogger
+}
+
+var _ storage.Backend = (*JetStreamStore)(nil)
+
+type JetStreamStoreOptions struct {
+	BucketPrefix string
+}
+
+type JetStreamStoreOption func(*JetStreamStoreOptions)
+
+func (o *JetStreamStoreOptions) apply(opts ...JetStreamStoreOption) {
+	for _, op := range opts {
+		op(o)
+	}
+}
+
+func NewJetStreamStore(ctx context.Context, conf *v1beta1.JetStreamStorageSpec, opts ...JetStreamStoreOption) (*JetStreamStore, error) {
+	options := JetStreamStoreOptions{
+		BucketPrefix: "gateway",
+	}
+	options.apply(opts...)
+
+	lg := logger.New(logger.WithLogLevel(zap.WarnLevel)).Named("jetstream")
+
+	nkeyOpt, err := nats.NkeyOptionFromSeed(conf.NkeySeedPath)
+	if err != nil {
+		return nil, err
+	}
+	nc, err := nats.Connect(conf.Endpoint, nkeyOpt, nats.MaxReconnects(-1))
```

Comment:
May want to add a custom disconnect handler e.g:
```
nats.DisconnectErrHandler(
	func(nc *nats.Conn, err error) {
		lg.With(
			"err", err,
		).Warn("nats disconnected")
	},
),
```

Reply:
A `nats.CustomReconnectDelay(func(int) time.Duration)` might also be an idea to prevent spamming the nats cluster if something happens

Reply:
:+1: also added RetryOnFailedConnect(true) which looks like it is needed to enable retries on the initial connection.